Refactor qapply to use SlidingTransform and multiple dispatch

ellisonbg opened the pull request:
#### References to other Issues or PRs

Related to #28022
Replacement for #24349
Fixes #27526, #27213

#### Brief description of what is fixed or changed

This PR introduces a major refactoring of the quantum physics package's operator application system. The key changes include:

1. New SlidingTransform Algorithm: Extracted and generalized the transformation algorithm used in quantum expression canonicalization into a reusable slidingtransform.py module. This algorithm provides a systematic way to apply transformations to multiplication expressions using a sliding window approach, which is particularly useful for non-commutative quantum operators.
2. Complete qapply Refactoring: Extensively refactored the qapply function to use the new SlidingTransform algorithm, improving its ability to handle complex quantum expressions involving operators, states, tensor products, and other quantum constructs.
3. Sum/Integral handlings: Sums and integrals are now properly handled in qapply.
4. Multiple-dispatch: The refactor uses multipledispatch to provide an extensible and clean way of defining how qapply works on different types of expressions.
5. No expand: qapply no longer calls `.expand()` as it works, resulting in more efficient calculations and results that are more similar to how humans do these types of calculations.
6. Less recursion: While recursion is still used in the algorithms, it is used in fewer places, resulting in a more efficient approach.
7. Enhanced Test Coverage: Added comprehensive tests for both the new SlidingTransform functionality and the refactored qapply system, including edge cases and complex quantum expressions.
8. Bug Fixes and Improvements: Fixed various issues in quantum operator applications, particularly around Grover operators, spin systems, and TensorProduct handling.

The refactoring is mostly backward compatibility while providing a more robust and extensible foundation for quantum symbolic computation.

#### Other comments

The SlidingTransform algorithm operates in two phases:

- Unary Phase: Processes individual factors in multiplications
- Binary Phase: Applies binary transformations to adjacent pairs using a sliding window

This approach allows for better handling of non-commutative quantum operators and provides a foundation for future quantum computation enhancements.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->

- physics.quantum
  - Introduced new SlidingTransform algorithm for systematic processing of quantum multiplication expressions
  - Major refactoring of qapply function for improved operator application handling
  - Enhanced support for complex quantum expressions involving tensor products and non-commutative operators
  - Full handling of sums and integrals in qapply
  - The qapply function no longer calls `.expand()` while it works so some of the results may have a slightly different form.
  - Fixed various bugs in Grover operators, spin systems, and quantum state applications
  - Improved test coverage for quantum operator applications and edge cases

<!-- END RELEASE NOTES -->

ellisonbg committed:
quantum: Generalize the transformer to be reusable.

In a recent PR, I introduced a new transformation algorithm that is used to canonicalize Mul's of quantum expressions. This generalizes the underlying algorithm and puts it in its own module for reuse. This will be used to refactor qapply.

ellisonbg committed:
Refactor qapply

ellisonbg committed:
quantum: Improvements to SlidingTransform and qapply

ellisonbg committed:
quantum: spin tests mostly working.

ellisonbg committed:
qapply: fix qapply on TensorProduct

ellisonbg committed:
quantum: all spin tests working!!!

ellisonbg committed:
quantum: fix grover tests and mark one as slow

ellisonbg committed:
quantum: fix grover iteration

Added a call to .expand after each grover iteration to avoid lots of products of long sums in qapply calls.

ellisonbg committed:
quantum: update todos for qapply and fix whitespace

ellisonbg committed:
quantum: Add docstrings and handle dispatch ambiguities

ellisonbg committed:
quantum: improve grover operator application

ellisonbg committed:
quantum: finish implementing qapply edge cases

ellisonbg committed:
quantum: more comprehensive tests for slidingtransform.py

ellisonbg committed:
quantum: remove extra whitespace

ellisonbg committed:
quantum: clean up issues found by flake8

ellisonbg committed:
quantum: fix doctests for quantum package

ellisonbg committed:
quantum: Add more qapply tests and fix edge cases in slidingtransform

ellisonbg committed:
quantum: reorganize qapply handler registration order and expand test coverage

- Reorder `_qapply_unary` registrations: quantum handlers (Dagger, Density, TensorProduct) before generic handlers (Expr, numbers)
- Reorder `qapply_Mul.unary` registrations: quantum handlers (OuterProduct, Commutator, etc.) before generic handlers
- Reorder `qapply_Mul.binary` registrations: operator-state application handlers before distribution handlers
- Reorganize test structure with clearer section headers for unary vs binary handlers
- Add placeholder test functions for all binary handler combinations
- Fix test naming consistency: `test_issue_3044`, `test_issue_24158_ket_times_op`

ellisonbg committed:
quantum: remove unused imports

ellisonbg committed:
quantum: Ignore pyright operator issue in test_qapply.py

ellisonbg committed:
quantum: add test for operator*tensor product

ellisonbg committed:
quantum: fix binary handlers and add comprehensive qapply combination tests

Fix bug in 6 binary handler functions (_qapply_mul_binary_sum_add,
_qapply_mul_binary_add_sum, _qapply_mul_binary_integral_add,
_qapply_mul_binary_add_integral, _qapply_mul_binary_sum_integral,
_qapply_mul_binary_integral_sum) that were returning bare objects
instead of tuples. The sliding transform machinery expects handlers
to return tuples so that _split_cnc can iterate over the results.

Add comprehensive test coverage for all binary combinations of
Add, Sum, Integral, and Expr (operator powers) with quantum operators
and states:
- test_mul_binary_add_add: Add * Add distribution
- test_mul_binary_sum_add: Sum * Add creates Sum(Add(...))
- test_mul_binary_add_sum: Add * Sum creates Sum(Add(...))
- test_mul_binary_integral_add: Integral * Add
- test_mul_binary_add_integral: Add * Integral
- test_mul_binary_sum_integral: Sum * Integral
- test_mul_binary_integral_sum: Integral * Sum
- test_mul_binary_expr_add: operator power * Add
- test_mul_binary_expr_sum: operator power * Sum
- test_mul_binary_expr_integral: operator power * Integral
- test_mul_binary_sum_sum: Sum * Sum creates double sum
- test_mul_binary_op_wavefunction: operator * wavefunction

All tests verify both symbolic results and numerical evaluation where
applicable. Tests use bosonic operators with Fock states to provide
concrete, verifiable quantum mechanical examples.

ellisonbg committed:
quantum: adding test_qapply changes

ellisonbg committed:
quantum: add test_qapply.py to quality_unicode.py exceptions

ellisonbg committed:
quantum: add comprehensive docstrings and doctests to qapply module

- Replace technical module docstring with user-friendly description
- Add detailed qapply() function docstring with Parameters, Returns, Examples sections
- Include 9 diverse doctests covering basic to advanced usage:
  * Basic operator-state application
  * Bosonic operators
  * Distribution over sums
  * Operator powers
  * Commutators and tensor products
  * Options (ip_doit, dagger)
  * Complex nested expressions
- All doctests verified to pass

ellisonbg commented:
@oscarbenjamin @akhmerov this PR has the qapply refactor we discussed in the previous iteration of the PR (#28023). The biggest changes since you last looked at it are comprehensive testing. This is ready for final review.

Also @Matt-Ord may find this of interest.

Matt-Ord commented on sympy/physics/quantum/slidingtransform.py:
```diff
@@ -0,0 +1,283 @@
+    def from_args(self):
+        return self._from_args
+
+    def __call__(self, expr, **options):
```
expr should be type annotated as mul if it only accepts mul

Matt-Ord replied:
Return type should be annotated as sp.Expr

Matt-Ord commented on sympy/physics/quantum/slidingtransform.py:
```diff
@@ -0,0 +1,283 @@
+            output = []
+
+        nc_parts = tuple(input)
+        c_parts = tuple(c_parts)
```
If both transforms were None is this correct?

Matt-Ord commented on sympy/physics/quantum/qapply.py:
```diff
@@ -1,232 +1,400 @@
+    return sympify(expr)
+
+
+#-----------------------------------------------------------------------------
```
What information does this comment provide?

Matt-Ord reviewed (changes requested):
Left a few comments but I haven't yet looked at the tests. Liked the use of Dispatcher - it makes it a lot cleaner to read

Matt-Ord commented on sympy/physics/quantum/slidingtransform.py:
```diff
@@ -0,0 +1,283 @@
+    return c, nc
+
+
+class SlidingTransform(object):
```
Does it need to inherit from object?

Matt-Ord replied:
Yeah removing would be best - reading the code made me think it was inheriting from some other class called object

ellisonbg commented on sympy/physics/quantum/slidingtransform.py:
```diff
@@ -0,0 +1,283 @@
+    def from_args(self):
+        return self._from_args
+
+    def __call__(self, expr, **options):
```
In the quantum package, we have been using file names without the underscores.

ellisonbg committed:
quantum: add type annotations to slidingtransform.py

ellisonbg commented on sympy/physics/quantum/slidingtransform.py:
```diff
@@ -0,0 +1,283 @@
+
+class SlidingTransform(object):
+
+    def __init__(self, unary=None, binary=None, reverse=False, from_args=True):
```
Fixed

ellisonbg committed:
quantum: add comprehensive type annotations to qapply module

The changes add extensive type annotations to qapply.py and related typing improvements to slidingtransform.py, including:

- Creating a QApplyOptions TypedDict for type-safe options
- Adding type hints to all dispatcher handlers
- Adding proper imports for type checking
- Improving type safety throughout the qapply module

ellisonbg committed:
quantum: add test of slidingtransform when both transforms are None

ellisonbg committed:
quantum: Fix reference to QApplyOptions in docs

ellisonbg committed:
quantum: fix typing issues in transform.py

Matt-Ord committed:
Stricter types in SlidingTransform

Matt-Ord committed:
Fix imports

Matt-Ord commented:
@ellisonbg I've made a small change to the types [in this pr](https://github.com/ellisonbg/sympy/pull/2) so that qapply mul also gets kwargs type checked, other than that it looks good to me!

Matt-Ord committed:
Remove TODO:

Matt-Ord committed:
Fix types

Matt-Ord committed:
Move Unpack to TYPE_CHECKING

ellisonbg committed:
Merge pull request #2 from Matt-Ord/sliding-transform-ty

Sliding transform ty

ellisonbg committed:
quantum: address review comments about type annotations

ellisonbg commented:
Review comments addressed, tests running again
